### PR TITLE
Propagate Replicas 0 to NovaMetadata CR

### PR DESCRIPTION
**Pull request by gibizer**
The existing logic ignored all NovaMetadata CRs with Replicas=0. This
was added to not create NovaMetadata in cells where it is not requested at
all. However this logic means that if NovaMetadata is deployed with
Replica>0 then later changed to Replica=0 then such update is simply
ignored.

We are mixing two cases into replicas 0:

1. the service does not need to deployed at all
2. the service needs to be temporary scaled down to 0 and then scaled
back up later

For the 2) we can clearly use the replicas field. But for 1)
we should use a different field not to mix the two cases.

So the current PR introduces an Enabled field to the NovaMetadata to
be used for case 1). If Enabled is set to false then the NovaMetadata
CR will not be created.

This patch also ensure that the Enabled field is defaulted to true for
the top level NovaMetadata and defaulted to false for the cell level
NovaMetadata templates.

Also the config generation for the nova-metadata-api is changed to only
assume API DB access if deployed on the top level. This was probably a
latent bug that is now visible. So it is fixed here.

We have to handle two edge cases too:

1.  If a Nova or NovaCell is deployed without Metadata enabled, then a Metadata is created manually with the appropriate name, and then the Metadata is enabled in the parent then the parent controller should not try to adopt the manually create Metadata as at least the label selector of the StatefulSet in immutable. So the parent goes to error state until the NovaMetadata CR is manually deleted.

2. Symmetrically if the Metadata is disabled while there is a Metadata CR then if the CR is owned by the parent Nova or NovaCell then the CR is deleted. However if the CR is not owned by the parent then the CR will by untouched.

~Depends-on: #459~ (merged)
Depends-on: https://github.com/openstack-k8s-operators/openstack-operator/pull/440

### CODE REVIEW

**Comment by gibizer** [comment]
Built this into openstack-operator and deployed with the default sample:

```
$ oc get NovaNoVNCPRoxy
NAME                    NETWORKATTACHMENTS   STATUS   MESSAGE
nova-cell1-novncproxy                        True     Setup complete


$ oc get NovaMetadata -n openstack -o yaml | yq --yaml-output  '.items[] | .metadata.name, .spec.replicas'
nova-cell1-metadata
--- 0
--- nova-metadata
--- 1
...

```

**Comment by gibizer** [comment]
just rebased it on top of #459 but I still need to implement the Enabled field

**Comment by SeanMooney** [comment]
the kuttl failures are vaild by the way
case.go:366: resource StatefulSet:nova-kuttl-default/nova-kuttl-metadata: .spec.replicas: value mismatch, expected: 1 != actual: 0

**Comment by gibizer** [comment]
/test nova-operator-build-deploy-tempest

```
--> Creating resources with label build=nova-operator ...
    error: Post "https://api.oko-06.oooci.ccitredhat.com:6443/apis/build.openshift.io/v1/namespaces/openstack/buildconfigs": EOF
--> Failed
```

**Comment by gibizer** [comment]
/test nova-operator-build-deploy

```
oc get secret/osp-secret || oc kustomize /alabama/install_yamls/out/openstack/input | oc apply -f -
Error from server (NotFound): secrets "osp-secret" not found
Error from server: error when creating "STDIN": etcdserver: request timed out
make: *** [Makefile:480: input] Error 1
```

**Comment by gibizer** [comment]
from the kuttl test, it seems we have a problem with defaulting...

https://review.rdoproject.org/zuul/build/37af7c74d45b439fbb3051237bf21878/log/job-output.txt#1187
```
The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
```

**Comment by gibizer** [comment]
> the kuttl failures are vaild by the way case.go:366: resource StatefulSet:nova-kuttl-default/nova-kuttl-metadata: .spec.replicas: value mismatch, expected: 1 != actual: 0

It is valid in a sense that kuttl asserting the wrong behavior (metadata cannot be scaled to 0). And now that the scaling works kuttl needs to assert the good behavior. I changed the kuttl test now to do so.

**Comment by gibizer** [comment]
> > from the kuttl test, it seems we have a problem with defaulting...
> > https://review.rdoproject.org/zuul/build/37af7c74d45b439fbb3051237bf21878/log/job-output.txt#1187
> > ```
> > The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
> > ```
> 
> I cannot reproduce this locally. With the recent kuttl commit in this PR the tests are passing for me locally.

This is still happening in the CI so I guess something is missing there, maybe the CRD is not updated according to the current PR content.

**Comment by gibizer** [comment]
> > from the kuttl test, it seems we have a problem with defaulting...
> > https://review.rdoproject.org/zuul/build/37af7c74d45b439fbb3051237bf21878/log/job-output.txt#1187
> > ```
> > The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
> > ```
> 
> I cannot reproduce this locally. With the recent kuttl commit in this PR the tests are passing for me locally.

I built this locally into the recent openstack-operator. I will see if this reproduces the problem:
```
OPENSTACK_IMG=quay.io/gibi/openstack-operator-index:v0.0.200 make openstack
```

**Comment by gibizer** [comment]
> > > from the kuttl test, it seems we have a problem with defaulting...
> > > https://review.rdoproject.org/zuul/build/37af7c74d45b439fbb3051237bf21878/log/job-output.txt#1187
> > > ```
> > > The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
> > > ```
> > 
> > 
> > I cannot reproduce this locally. With the recent kuttl commit in this PR the tests are passing for me locally.
> 
> I built this locally into the recent openstack-operator. I will see if this reproduces the problem:
> 
> ```
> OPENSTACK_IMG=quay.io/gibi/openstack-operator-index:v0.0.200 make openstack
> ```

With that build + the nova kuttl test from this PR I can reproduce the issue:
```
$ PATH="/home/gibi/bin/sh-to-bash:$PATH" make kuttl-test
oc apply -k test/kuttl/test-suites/default/deps/ --timeout=120s
namespace/nova-kuttl-default created
secret/osp-secret created
The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
make: *** [Makefile:364: kuttl-test-prep] Error 1
```

**Comment by gibizer** [comment]
> > > > from the kuttl test, it seems we have a problem with defaulting...
> > > > https://review.rdoproject.org/zuul/build/37af7c74d45b439fbb3051237bf21878/log/job-output.txt#1187
> > > > ```
> > > > The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
> > > > ```
> > > 
> > > 
> > > I cannot reproduce this locally. With the recent kuttl commit in this PR the tests are passing for me locally.
> > 
> > 
> > I built this locally into the recent openstack-operator. I will see if this reproduces the problem:
> > ```
> > OPENSTACK_IMG=quay.io/gibi/openstack-operator-index:v0.0.200 make openstack
> > ```
> 
> With that build + the nova kuttl test from this PR I can reproduce the issue:
> 
> ```
> $ PATH="/home/gibi/bin/sh-to-bash:$PATH" make kuttl-test
> oc apply -k test/kuttl/test-suites/default/deps/ --timeout=120s
> namespace/nova-kuttl-default created
> secret/osp-secret created
> The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
> make: *** [Makefile:364: kuttl-test-prep] Error 1
> ```

And it turned out it is a bug in openstack-operator. Fix is proposed https://github.com/openstack-k8s-operators/openstack-operator/pull/440

**Comment by gibizer** [comment]
> > With that build + the nova kuttl test from this PR I can reproduce the issue:
> > ```
> > $ PATH="/home/gibi/bin/sh-to-bash:$PATH" make kuttl-test
> > oc apply -k test/kuttl/test-suites/default/deps/ --timeout=120s
> > namespace/nova-kuttl-default created
> > secret/osp-secret created
> > The OpenStackControlPlane "openstack" is invalid: spec.nova.template.metadataServiceTemplate.enabled: Invalid value: "null": spec.nova.template.metadataServiceTemplate.enabled in body must be of type boolean: "null"
> > make: *** [Makefile:364: kuttl-test-prep] Error 1
> > ```
> 
> And it turned out it is a bug in openstack-operator. Fix is proposed [openstack-k8s-operators/openstack-operator#440](https://github.com/openstack-k8s-operators/openstack-operator/pull/440)

I've built #440 and this PR together into `OPENSTACK_IMG=quay.io/gibi/openstack-operator-index:v0.0.201 make openstack_wait` and now kuttl passing on it.

**Comment by gibizer** [comment]
> just rebased it on top of #459 but I still need to implement the Enabled field

Now the enabled field is fully implemented

**Comment by gibizer** [comment]
recheck fix in https://github.com/openstack-k8s-operators/openstack-operator/pull/440 merged

**Comment by SeanMooney** [comment]
recheck

**Comment by gibizer** [comment]
/unhold
parent PR merged

**Comment by gibizer** [comment]
It just needed a trivial rebase

**Comment by gibizer** [comment]
recheck
zuul based CI issue is fixed

File: api/v1beta1/nova_webhook.go [hunk 1]
```diff
@@ -83,6 +83,13 @@ func (spec *NovaSpec) Default() {
 		spec.MetadataServiceTemplate.ContainerImage = novaDefaults.MetadataContainerImageURL
 	}
 
+	// NOTE(gibi): this cannot be expressed kubebuilder defaults as the
```
Comment (SeanMooney) [inline api/v1beta1/nova_webhook.go:86]: nit: 
```
this cannot be expressed as kubebuilder ...
```
or
```
this cannot be expressed via kubebuilder ...
```

File: test/functional/nova_multicell_test.go [hunk 2]
```diff
@@ -803,6 +800,24 @@ var _ = Describe("Nova multicell", func() {
...
+			th.SimulateStatefulSetReplicaReady(cell0.ConductorStatefulSetName)
+			th.SimulateJobSuccess(cell0.CellMappingJobName)
+
+			cell1Secret := th.GetSecret(cell1.InternalCellSecretName)
+			Expect(cell1Secret.Data).To(
+				HaveKeyWithValue(controllers.MetadataSecretSelector, []byte("metadata-secret")))
+			cell0Secret := th.GetSecret(cell0.InternalCellSecretName)
+			Expect(cell0Secret.Data).NotTo(
+				HaveKeyWithValue(controllers.MetadataSecretSelector, []byte("metadata-secret")))
 		})
```
Comment (SeanMooney) [inline test/functional/nova_multicell_test.go:821]: +1

File: test/functional/nova_reconfiguration_test.go [hunk 3]
```diff
@@ -510,4 +511,70 @@ var _ = Describe("Nova reconfiguration", func() {
...
+		}, timeout, interval).Should(Succeed())
+
+		AssertMetadataDoesNotExist(novaNames.MetadataName)
+		th.ExpectCondition(
+			novaNames.NovaName,
+			ConditionGetterFunc(NovaConditionGetter),
+			condition.ReadyCondition,
+			corev1.ConditionTrue,
+		)
+		nova := GetNova(novaNames.NovaName)
```
Comment (SeanMooney) [inline test/functional/nova_reconfiguration_test.go:546]: i was going to suggest just creating the nova CR with it disabled then i realsed this si the reconfiguration tests
so ya this is fine +1

**Review by SeanMooney — APPROVED** [review]
i have 1 nit inline but i dont think we need to hold this for that so approving.
thanks gibi

File: api/v1beta1/nova_webhook.go (hunk 1 above)
Comment (gibizer) [reply api/v1beta1/nova_webhook.go:86]: followed up in https://github.com/openstack-k8s-operators/nova-operator/pull/517

**Pull request merged**